src/python/pants/rules/core/list_roots.py

```diff
@@ -0,0 +1,61 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from typing import Set
+
+from pants.engine.console import Console
+from pants.engine.fs import PathGlobs, Snapshot
+from pants.engine.goal import Goal, LineOriented
+from pants.engine.rules import console_rule, optionable_rule, rule
+from pants.engine.selectors import Get
+from pants.source.source_root import AllSourceRoots, SourceRoot, SourceRootConfig
+
+
+class Roots(LineOriented, Goal):
+  """List the repo's registered source roots."""
+  name = 'roots'
+
+
+@rule(AllSourceRoots, [SourceRootConfig])
+def all_roots(source_root_config):
+
+  source_roots = source_root_config.get_source_roots()
+
+  all_paths: Set[str] = set()
+  for path in source_roots.traverse():
+    if path.startswith("^/"):
+      all_paths.add(f"{path[2:]}/")
+    else:
+      all_paths.add(f"**/{path}/")
+
+  snapshot = yield Get(Snapshot, PathGlobs(include=tuple(all_paths)))
+
+  all_source_roots: Set[SourceRoot] = set()
+
+  # The globs above can match on subdirectories of the source roots.
+  # For instance, `src/*/` might mach 'src/rust/' as well as
```

Comment:
s/mach/match/

Reply:
Otherwise, perfect comment! Very clear.

Reply:
Oh, but let's add a test for exactly this example, so anyone reading this comment can see it in action.

Reply:
*smacks forehead* will fix typo and add tests